### PR TITLE
Lora extraction

### PR DESCRIPTION
**Pull request by thomasgauthier**
This PR adds support for extracting PEFT compatible LoRAs out of fine tune models as demonstrated in my [LoRD](https://github.com/thomasgauthier/LoRD/) repo.

Potential improvements:
  - Automatic model card creation (right now the script just outputs config and weights)
  - Better heuristics around choosing a proper alpha value. The current version uses alpha=rank as it seemed to give good result empirically but there might be some other trick we could use to find good alpha values (gradient based or otherwise)
  - Evals and metrics to assess how viable this technique is
  - Ability to use a custom rank for each weight tensor, some weight matrices might be contributing more to the task vectors and might benefit from higher ranks
    - Automating the search for good rank values seems like a promising area of research, [Fernando Neto](https://twitter.com/FernandoNetoAi) shared some interesting ideas with me for accomplishing this using Random Matrix Theory. See my comments below

### CODE REVIEW

**Comment by thomasgauthier** [comment]
Optimizing rank value using Random Matrix Theory, GPT-4 writeup
---

The discussion revolves around an innovative approach to optimizing the adaptation of fine-tuned models using principles from Random Matrix Theory (RMT) and Singular Value Decomposition (SVD). This method seeks to refine how low-rank adapters (matrices \(A\) and \(B\)) are extracted from fine-tuned models, with the aim of determining the optimal rank for these adapters in a more principled and theoretically grounded way. Here's a breakdown of the approach and the ideas presented:

### Conceptual Foundation

- **Fine-tuning as a Delta**: In the context of model adaptation, fine-tuning a pre-trained model for a specific task can be seen as applying a delta (\(\Delta W\)) to the model's original weights (\(W\)). This delta represents the changes or adjustments necessary to tailor the model's behavior to the new task.
  
- **Extraction of Low-Rank Adapters**: The approach proposes extracting low-rank adapters (\(A\) and \(B\)) from this delta using Singular Value Decomposition (SVD). SVD is a mathematical method that decomposes a matrix into three other matrices, where \(A\) and \(B\) can be seen as representing the most significant directions of variation (or information) in \(\Delta W\), constrained by a given rank.

### Application of Random Matrix Theory (RMT)

- **Assessment of Residuals**: The key innovation here involves using RMT to assess the residuals resulting from the difference between the delta (\(\Delta W\)) and the product of the low-rank adapters (\(A \times B\)). The residuals (\(\Delta W - A \times B\)) represent the portion of the delta not captured by the low-rank approximation.

- **Noise Detection in Residuals**: RMT is employed to determine whether these residuals are essentially noisy or contain structured information. The assumption is that if the residuals are mostly noise, then the chosen rank for \(A\) and \(B\) is sufficiently capturing the significant information in \(\Delta W\). Conversely, if the residuals contain structured information, it might indicate that the rank is too low and important variations are being lost.

### Determining the Optimal Rank

- **Rank Dependency on Orthogonality**: The optimal rank for the adapters depends on various factors, including how different (orthogonal) two models are from each other. The idea is that the more orthogonal the changes introduced by fine-tuning, the higher the potential rank needed to capture these differences effectively without resorting to arbitrarily high or low ranks.

- **Challenges with Arbitrary Rank Selection**: Selecting a rank that is too low might miss significant information, while a rank that is too high might capture noise as if it were important, leading to inefficiencies and potential overfitting.

- **Unifying Approaches**: The proposition is to unify the application of SVD for extracting low-rank adapters with the insights provided by RMT on the nature of residuals. This unified approach aims to guide the selection of an optimal rank for \(A\) and \(B\) in a way that is both principled and responsive to the specific characteristics of the models and tasks involved.

### Conclusion

This approach offers a sophisticated method for determining the optimal rank for low-rank adapters in model fine-tuning, leveraging the strengths of SVD for matrix decomposition and RMT for evaluating the informational content of residuals. By doing so, it seeks to provide a more effective and theoretically informed pathway to model optimization, reducing the guesswork and inefficiencies associated with arbitrary rank selection. This method not only enhances the efficiency and performance of the adapted models but also contributes to a deeper understanding of the underlying dynamics of model fine-tuning and adaptation.

---

Note: this idea isn't mine but was shared with me by [Fernando Neto](https://twitter.com/FernandoNetoAi). He and other collaborators used similar ideas in their [LASER implementation](https://github.com/tdolan21/laserRMT)

File: mergekit/scripts/extract_lora.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+from safetensors.torch import save_file
+from tqdm import tqdm
+from transformers import AutoConfig, AutoModelForCausalLM
+
+from mergekit.common import ModelReference
+from mergekit.io import LazyTensorLoader
+from mergekit.options import add_merge_options
+
+
+def _low_rank_decomposition(weight, reduced_rank=16):
```
Comment (cg123) [inline mergekit/scripts/extract_lora.py:17]: A little nitpicky but it would be great to have type annotations on these helper functions.

File: mergekit/scripts/extract_lora.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        ):
+            names.append(name)
+
+    return names
+
+
+def get_linear_module_names(model_id):
+    model = AutoModelForCausalLM.from_pretrained(
+        model_id, state_dict={}, device_map="meta"
+    )  # avoid loading weights as we won't need them
```
Comment (cg123) [inline mergekit/scripts/extract_lora.py:85]: 👍 Good approach for this. Someday we might put all of the info needed for this in ArchitectureInfo but this is perfect right now.

File: mergekit/scripts/extract_lora.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        lora_weights[f"base_model.model.{layer_name}.lora_A.weight"] = lora_A.to(
+            "cpu"
+        ).contiguous()
+        lora_weights[f"base_model.model.{layer_name}.lora_B.weight"] = lora_B.to(
+            "cpu"
+        ).contiguous()
+
+    lora_config = create_peft_config(
+        base_model_name_or_path=base_model_ref.model.path,
+        alpha=rank,  # Setting the alpha to the reduced rank value (instead of alpha value used) seems to give better performance. Further testing would be needed to understand what is the optimal alpha value to use
```
Comment (cg123) [inline mergekit/scripts/extract_lora.py:178]: This is correct yeah! The delta is scaled by `alpha/r` (see [here](https://github.com/huggingface/peft/blob/02b5aeddf9c1ea11451f10a8a26da7e5df8cca4a/src/peft/tuners/lora/layer.py#L107)) so `alpha=r` is going to give you the closest possible approximation of the fine tuned model when applied to the base.

File: mergekit/scripts/extract_lora.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    linear_module_names = get_linear_module_names(base_model_ref.model.path)
+    finetuned_model_linear_module_names = get_linear_module_names(
+        finetuned_model_ref.model.path
+    )
+
+    assert set(linear_module_names) == set(
+        finetuned_model_linear_module_names
+    ), "Model architecture mismatch"
+
+    base_loader = LazyTensorLoader(base_model_ref.tensor_index(), lazy_unpickle=True)
```
Comment (cg123) [inline mergekit/scripts/extract_lora.py:155]: It'd be good to add an option for `--lazy-unpickle/--no-lazy-unpickle` for the occasional weird file that the lazy unpickler can't handle.

**Review by cg123 — COMMENTED** [review]
Looking pretty good! Two little nitpicks on my end - looking forward to getting this in.

**Comment by dmahurin** [comment]
The 4 lora-extraction commits rebased on current main, if it helps.

https://github.com/dmahurin/mergekit/tree/lora-extraction

**Comment by cg123** [comment]
Looks great, let's get this in there!
